From 9887790ee13cac81130e29ed31c0c095839da858 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Mon, 28 Aug 2017 11:36:13 +0530
Subject: [PATCH 353/532] v4l: xilinx: sdirxss: Support decoding ST352 payload
 for 12G mode

commit 3c5133ada3046ef81602fb8e048cc0fe4eea48e8 from
https://github.com/Xilinx/linux-xlnx.git

Patch adds support for decoding ST352 payload for 12G modes to identify
if resolution is 4096x2160 or 3840x2160.

Signed-off-by: Vishal Sagar <vsagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-sdirxss.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index 42d8df4..6f615be 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -865,9 +865,27 @@ static int xsdirxss_get_format(struct v4l2_subdev *sd,
 	case XSDIRX_MODE_6G_MASK:
 		break;
 	case XSDIRX_MODE_12GI_MASK:
-		break;
 	case XSDIRX_MODE_12GF_MASK:
+		switch (payload & 0xFF) {
+		case 0xCE:
+			/* Section 4.3.1 SMPTE ST 2082-10 */
+			fmt->format.height = 2160;
+			if (payload & 0x00400000)
+				/*
+				 * bit 6 of byte 3 indicates whether
+				 * 4096 (1) or 3840 (0)
+				 */
+				fmt->format.width = 4096;
+			else
+				fmt->format.width = 3840;
+			break;
+		default:
+			dev_dbg(core->dev, "Unknown 12G Mode SMPTE standard\n");
+		};
 		break;
+	default:
+		dev_dbg(core->dev, "Invalid Mode\n");
+		return -EINVAL;
 	}
 
 	if (payload & 0x4000)
-- 
2.7.4

